src/charm.py

```diff
@@ -199,6 +226,61 @@ def _restart(self, event: EventBase):
 
         self.container.restart(CHARM_KEY)
 
+        # Indicate that unit has completed restart
+        self.cluster.relation.data[self.unit]["password-rotated"] = "true"
+
+        # If leader runs last on RollingOps restart, this code would be enough
+        """
+        if self.unit.is_leader() and self.cluster.relation.data[self.app].get("rotate-passwords"):
+            logger.error("LEADER REMOVING ROTATE_PASSWORDS")
+            self.cluster.relation.data[self.app]["rotate-passwords"] = ""
+        """
+
+    def _get_super_password_action(self, event: ActionEvent) -> None:
+        """Handler for get-super-password action event."""
+        event.set_results({"super-password": self.cluster.passwords[0]})
+
+    def _get_sync_password_action(self, event: ActionEvent) -> None:
+        """Handler for get-sync-password action event."""
+        event.set_results({"sync-password": self.cluster.passwords[1]})
+
+    def _set_password_action(self, event: ActionEvent) -> None:
+        """Handler for set-password action.
+
+        Set the password for a specific user, if no passwords are passed, generate them.
+        """
+        if not self.unit.is_leader():
+            msg = "Password rotation must be called on leader unit"
+            logger.error(msg)
+            event.fail(msg)
+            return
+
+        username = "super"
+        if "username" in event.params:
+            username = event.params["username"]
+        if username not in CHARM_USERS:
+            msg = f"The action can be run only for users used by the charm: {CHARM_USERS} not {username}."
+            logger.error(msg)
+            event.fail(msg)
+            return
+
+        new_password = self.cluster.generate_password()
+        if "password" in event.params:
+            new_password = event.params["password"]
+
+        # Passwords should not be the same.
+        if new_password in self.cluster.passwords:
+            event.log("The old and new passwords are equal.")
+            event.set_results({f"{username}-password": new_password})
+            return
```

Comment:
This is more for user provided passwords, rather than generated ones. This will get updated in the future though, as I'll add a bit of logic to check password strength.